### ISSUE TITLE
Supervised BERTopic approach advice

### ISSUE THREAD

**Issue by jasonmorkel** [MaartenGr/BERTopic#511]
Hi Maarten, 

Hope you are well. 

My dataset consists of free text survey responses to 10 different questions in a survey and I would like to perform topic modelling on the responses. 

Here are the labels I gave to the first 4 questions. 

`['Course assessments: Comments', 'Course Content: Comments',
       'Enhancements to OLC: Comments', 'Overall experience: Comments',`

My question to you is- would it be best to perform a normal unsupervised BERTopic model on each question or would it be better to perform a supervised BERTopic model using the question names as categories?

**Comment by MaartenGr** [comment]
I am! Thank you for asking. 

This is a bit of an annoying answer but it "depends". I would not say that one is necessarily better than the other as it depends mostly on your use case and the labels that you have. 

In your case, assuming you have sufficient data and labels, it might be worthwhile to use the labels especially if you know they are quite accurate. That way, it is a bit easier to find topics within topics (i.e., labels) and allows you to dive a bit deeper into the classes you have already defined. A downside to doing this is that you are nudging the topic creation a bit towards those specific classes. However, what if we can find topics that overlap several classes? There might be a chance, albeit small, that those overlapping topics will not be found. 

In general, (semi)-supervised topic modeling in BERTopic is mostly about nudging the topic creation one way or another and is typically not fully supervised. In practice, if possible, I would advise performing both and see whether you notice any differences when doing so. If for whatever reason, you can only choose one method, then I would just stay with the unsupervised approach if you are generally looking for topics or choose the semi-supervised approach if you want to dig a little deeper into those classes. 

Hopefully, this (non-)answer helps a bit!

**Comment by jasonmorkel** [comment]
Hi Maarten, 

Thanks for your swift reply. 

I guess with these kind of tools, as you say, it comes down to the specific use case and the material that you are working with for the project. 

You have had some valid points and the best thing to do in this situation is just to not be lazy and try out all avenues. 

In my case, the questions in the surveys are quite distinct and there is a fair amount of data for the project (250,000+). We want to dig a bit deeper in each class so I ran the semi-supervised approach and it yielded fantastic results. 

The only issue is that there are quite a few responses that are being labelled as -1 (outliers). 

Here is the code that I used: 

`from bertopic import BERTopic
from sklearn.feature_extraction.text import CountVectorizer
vectorizer_model = CountVectorizer(ngram_range=(1, 3), stop_words="english")

topic_model = BERTopic(vectorizer_model=vectorizer_model, min_topic_size = 50, calculate_probabilities=True, verbose=True).fit(docs, y=categories)`

The number of topics created is 330 and +-45% of survey responses are being assigned to -1. 

I looked at some of the responses being assigned to -1 and after looking at the topics created, I would think that some of them should definitely be labelled a certain topic.

Would you suggest the only method to fix this would be to assign outliers to clusters only if they exceed a threshold probability. 

Something like this? 

`probability_threshold = 0.01
new_topics = [np.argmax(prob) if max(prob) >= probability_threshold else -1 for prob in probs]`

**Comment by jasonmorkel** [comment]
Hi @MaartenGr, 

Thanks again for your elaborate explanations! 

So I have played around a bit with modifying the hdbscan_model parameters but for my scenario, it doesn't make that much of a difference. 

I think the best approach would be to classify the outliers using the probability threshold approach. I will play around with the threshold so that only 5%-10% of the data is left as outliers. 

I think the outliers can also give valuable insights especially when applying sentiment so for my use case its good to have a few! 

Just to confirm one thing with you, if I run:
`probability_threshold = 0.01`
`new_topics = [np.argmax(prob) if max(prob) >= probability_threshold else -1 for prob in probs]`

Then you say there is a chance that non-outlier topics might be clustered differently? So it would be best to write code that only cluster the -1 outliers into non-outlier topics. 

> Lastly, it might be an interesting experiment to look at the upcoming https://github.com/MaartenGr/BERTopic/pull/492 (or v0.10.0, not sure yet about the version) release of BERTopic in which you can use any other clustering technique, like k-Means. You can install that PR and run it like so:

Amazing work regarding the upgrade! The additional functionality will be great. When do you plan on releasing it?

**Comment by MaartenGr** [comment]
> So I have played around a bit with modifying the hdbscan_model parameters but for my scenario, it doesn't make that much of a difference.

That is a shame, I was hoping that would have been a quick fix!

> Then you say there is a chance that non-outlier topics might be clustered differently? So it would be best to write code that only cluster the -1 outliers into non-outlier topics.

Yes! It should not be that significant but I believe it would be best to only classify the -1 outliers into non-outlier topics. 

> Amazing work regarding the upgrade! The additional functionality will be great. When do you plan on releasing it?

Fortunately, I only need to do some restructuring/cleaning/documentation and no other features, so I expect it to be released within a week or two.

**Comment by jasonmorkel** [comment]
Amazing work, thanks Maarten!
